torch/_tensor_str.py

```diff
@@ -65,132 +65,130 @@ def set_printoptions(
-    # TODO: use fmod?
-    for value in tensor.tolist():
-        if value != math.ceil(value):
-            int_mode = False
-            break
-
-    exp_min = tensor.min()
-    if exp_min != 0:
-        exp_min = math.floor(math.log10(exp_min)) + 1
-    else:
-        exp_min = 1
-    exp_max = tensor.max()
-    if exp_max != 0:
-        exp_max = math.floor(math.log10(exp_max)) + 1
-    else:
-        exp_max = 1
-    include_decimal_int_mode = floating_dtype and int_mode
-
-    scale = 1
-    exp_max = int(exp_max)
-    prec = PRINT_OPTS.precision
-    if int_mode:
-        if exp_max > prec + 1:
-            format = '{{:11.{}e}}'.format(prec)
-            fmt_fn = format.format
-            sz = max(min_sz, 7 + prec)
-        else:
-            sz = max(min_sz, exp_max + 1 + include_decimal_int_mode)
-            format = '{:' + str(sz) + '.0f}'
-            fmt_fn = format.format
-            if include_decimal_int_mode:
-                format = '{:' + str(sz - 1) + '.0f}'
-                nonfinite_format = format + '.'
-                fmt_fn = _get_format_fn(format, nonfinite_format)
-    else:
-        if exp_max - exp_min > prec:
-            sz = 7 + prec
-            if abs(exp_max) > 99 or abs(exp_min) > 99:
-                sz = sz + 1
-            sz = max(min_sz, sz)
-            format = '{{:{}.{}e}}'.format(sz, prec)
-            fmt_fn = format.format
-        else:
-            if exp_max > prec + 1 or exp_max < 0:
-                sz = max(min_sz, 7)
-                scale = math.pow(10, max(exp_max - 1, _min_log_scale))
-            else:
-                if exp_max == 0:
-                    sz = 7
+                if exp_max - exp_min > PRINT_OPTS.precision or exp_max > 8 or exp_min < -4:
+                    self.sci_mode = True
+                    for value in copy_list:
+                        value_str = ('{:.' + str(PRINT_OPTS.precision) + 'e}').format(value)
+                        self.max_width = max(self.max_width, len(value_str))
                 else:
-                    sz = exp_max + 6
-                sz = max(min_sz, sz)
-            format = '{{:{}.{}f}}'.format(sz, prec)
-            fmt_fn = format.format
-    return fmt_fn, scale, sz
+                    for value in copy_list:
+                        value_str = ('{:.' + str(PRINT_OPTS.precision) + 'f}').format(value)
+                        self.max_width = max(self.max_width, len(value_str))
+
+    def width(self):
+        return self.max_width
+
+    def format(self, value):
+        if self.floating_dtype:
+            if self.int_mode:
+                ret = '{:.0f}'.format(value)
+                if not (math.isinf(value) or math.isnan(value)):
+                    ret += '.'
+            elif self.sci_mode:
+                ret = ('{:' + str(self.max_width) +
+                       '.' + str(PRINT_OPTS.precision) + 'e}').format(value)
+            else:
+                ret = ('{:.' + str(PRINT_OPTS.precision) + 'f}').format(value)
```

Comment:
nit: you have the actual formats multiple times -- can you have them call into a common function?

Reply:
Actually only 2 of the 4 are the same, but I'll change the concat to a format to make it clearer.